In [8]:
from pathlib import Path
import pandas as pd

# Display Setting
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)

# Dataset Path
DATA_PATH = Path("../data/raw")


### Import Dataset


In [9]:
customers = pd.read_csv(DATA_PATH / "olist_customers_dataset.csv")
orders = pd.read_csv(DATA_PATH / "olist_orders_dataset.csv")
order_items = pd.read_csv(DATA_PATH / "olist_order_items_dataset.csv")
payments = pd.read_csv(DATA_PATH / "olist_order_payments_dataset.csv")
products = pd.read_csv(DATA_PATH / "olist_products_dataset.csv")
reviews = pd.read_csv(DATA_PATH / "olist_order_reviews_dataset.csv")
sellers = pd.read_csv(DATA_PATH / "olist_sellers_dataset.csv")
geo = pd.read_csv(DATA_PATH / "olist_geolocation_dataset.csv")
translation = pd.read_csv(DATA_PATH / "product_category_name_translation.csv")

### Dataset Overview

Pada bagian ini dilakukan identifikasi awal terhadap seluruh dataset yang digunakan dalam project **InsightFlow**.

Tujuan dari tahap ini adalah untuk memahami cakupan dataset, jumlah tabel yang tersedia, serta ukuran masing-masing dataset sebelum dilakukan analisis lebih lanjut.

In [10]:
datasets = {
    "Customers": customers,
    "Orders": orders,
    "Order Items": order_items,
    "Payments": payments,
    "Products": products,
    "Reviews": reviews,
    "Sellers": sellers,
    "Geolocation": geo,
    "Category Translation": translation,
}

In [11]:
overview = pd.DataFrame([
    {
        "Dataset": name,
        "Rows": df.shape[0],
        "Columns": df.shape[1]
    }
    for name, df in datasets.items()
])

overview

,Dataset,Rows,Columns
0,Customers,99441,5
1,Orders,99441,8
2,Order Items,112650,7
3,Payments,103886,5
4,Products,32951,9
5,Reviews,99224,7
6,Sellers,3095,4
7,Geolocation,1000163,5
8,Category Translation,71,2


### File Inventory
Bagian ini mendokumentasikan seluruh file yang tersedia pada dataset Olist beserta fungsi utamanya.

Seluruh dataset pada tahap ini masih berstatus **Available**, yang berarti seluruh file akan dipertimbangkan terlebih dahulu sebelum diputuskan apakah digunakan dalam MVP atau tidak.

Keputusan akhir mengenai penggunaan setiap dataset akan dilakukan pada tahap **Business Mapping** dan **Dataset Assessment**.

In [12]:
inventory = pd.DataFrame({
    "Dataset":[
        "Customers",
        "Orders",
        "Order Items",
        "Products",
        "Payments",
        "Reviews",
        "Sellers",
        "Geolocation",
        "Category Translation"
    ],
    "Primary Purpose":[
        "Customer Information",
        "Order Information",
        "Order Details",
        "Product Information",
        "Payment Information",
        "Customer Reviews",
        "Seller Information",
        "Geographic Reference",
        "Category Translation"
    ],
    "Status":[
        "Available",
        "Available",
        "Available",
        "Available",
        "Available",
        "Available",
        "Available",
        "Available",
        "Available"
    ]
})

inventory

,Dataset,Primary Purpose,Status
0,Customers,Customer Information,Available
1,Orders,Order Information,Available
2,Order Items,Order Details,Available
3,Products,Product Information,Available
4,Payments,Payment Information,Available
5,Reviews,Customer Reviews,Available
6,Sellers,Seller Information,Available
7,Geolocation,Geographic Reference,Available
8,Category Translation,Category Translation,Available


## Customers Dataset


In [16]:
customers.head()

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP


In [17]:
customers.sample(5)

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
21061,90460028374bf093ad66cb3722b67b14,9eb4ca952e7e366483ff31a07619cad5,37640,extrema,MG
66238,ba9e2ff94ab776d05be0cbd067755c50,a73e7aa41ad2bb8f870bbe407ab500cf,11960,eldorado,SP
25404,b5cfb32c2e07f8782e9c3d5e9e18918b,c1ed8887378ed3cee44474bd42508438,7600,mairipora,SP
78595,6093fdd0645070ffa0c013381b6cf31b,116599835eb984c240446753380232c1,36300,sao joao del rei,MG
12057,c6effe6fa35c9e0cbd5e9cf161fa2389,ffb2ae2aa0951d0885d7cd9a218b0c46,23058,rio de janeiro,RJ


In [19]:
customers.shape

(99441, 5)

In [20]:
customers.info()

<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype
---  ------                    --------------  -----
 0   customer_id               99441 non-null  str  
 1   customer_unique_id        99441 non-null  str  
 2   customer_zip_code_prefix  99441 non-null  int64
 3   customer_city             99441 non-null  str  
 4   customer_state            99441 non-null  str  
dtypes: int64(1), str(4)
memory usage: 3.8 MB


In [21]:
customers.describe(include="all")

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
count,99441,99441,99441.000000,99441,99441
unique,99441,96096,NaN,4119,27
top,06b8999e2fba1a1fbc88172c00ba8bc7,8d50f5eadf50201ccdcedfb9e2ac8455,NaN,sao paulo,SP
freq,1,17,NaN,15540,41746
mean,NaN,NaN,35137.474583,NaN,NaN
std,NaN,NaN,29797.938996,NaN,NaN
min,NaN,NaN,1003.000000,NaN,NaN
25%,NaN,NaN,11347.000000,NaN,NaN
50%,NaN,NaN,24416.000000,NaN,NaN
75%,NaN,NaN,58900.000000,NaN,NaN


In [22]:
customers.isna().sum()

customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

In [23]:
customers.duplicated().sum()

np.int64(0)

In [24]:
customers.nunique()

customer_id                 99441
customer_unique_id          96096
customer_zip_code_prefix    14994
customer_city                4119
customer_state                 27
dtype: int64

#### Initial Observation Customers Dataset

- Dataset terdiri dari **99,441** customer records dan **5** kolom.
- Tidak ditemukan missing values pada seluruh kolom.
- Tidak ditemukan baris duplikat.
- Terdapat 4 Kolom dengan data type `str` dan 1 Kolom dengan data type `int64`
- `customer_id` bersifat unik per baris (99,441 unique = total baris)  namun ini adalah identifier di level order/transaksi, bukan identifier pelanggan asli.
- `customer_unique_id` TIDAK unik per baris (96,096 unique dari 99,441 baris) — terdapat 3,345 baris dengan customer_unique_id yang berulang, menunjukkan adanya pelanggan yang melakukan lebih dari satu order.
- Kesimpulan: `customer_unique_id` adalah identifier pelanggan sesungguhnya dan WAJIB digunakan untuk seluruh analisis customer-level. `customer_id` hanya valid digunakan sebagai identifier order.

## Orders Dataset

In [25]:
orders.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


In [26]:
orders.sample(5)

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
79847,6bcce39be127575854779e39a50f2059,d8b65951c84554c80270720c9b5f1178,delivered,2018-04-28 02:23:24,2018-05-01 05:32:22,2018-05-04 11:35:00,2018-05-08 00:18:58,2018-05-17 00:00:00
38734,3e24f7d95766568a604655166cde923a,1a5dbe17955f253dba47bb5853716345,delivered,2017-05-24 15:55:43,2017-05-24 16:05:25,2017-05-26 07:04:58,2017-06-01 16:45:45,2017-06-19 00:00:00
3938,a5df4819e2227684e86dc2959b9ae9d6,fd6201ea30defd9bcee3bb0e09c6d2c9,delivered,2018-08-01 23:28:16,2018-08-01 23:43:52,2018-08-03 10:37:00,2018-08-08 19:48:03,2018-08-20 00:00:00
30312,5e82591b43cb83a7dd419ff564e44d04,af7177e092a630042e08c4301b7421bd,delivered,2017-09-18 16:17:16,2017-09-18 16:35:37,2017-09-19 14:44:18,2017-09-27 19:45:09,2017-10-09 00:00:00
61554,ab9a3c29c4b67aaca311cad95058153a,84c25336ff90f2e2b233e6df15840570,delivered,2017-05-11 13:56:39,2017-05-11 14:10:18,2017-05-12 12:43:00,2017-05-15 07:13:50,2017-05-19 00:00:00


In [27]:
orders.shape

(99441, 8)

In [28]:
orders.info()

<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype
---  ------                         --------------  -----
 0   order_id                       99441 non-null  str  
 1   customer_id                    99441 non-null  str  
 2   order_status                   99441 non-null  str  
 3   order_purchase_timestamp       99441 non-null  str  
 4   order_approved_at              99281 non-null  str  
 5   order_delivered_carrier_date   97658 non-null  str  
 6   order_delivered_customer_date  96476 non-null  str  
 7   order_estimated_delivery_date  99441 non-null  str  
dtypes: str(8)
memory usage: 6.1 MB


In [29]:
orders.describe(include="all")

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
count,99441,99441,99441,99441,99281,97658,96476,99441
unique,99441,99441,8,98875,90733,81018,95664,459
top,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2018-03-31 15:08:21,2018-02-27 04:31:10,2018-05-09 15:48:00,2018-05-14 20:02:44,2017-12-20 00:00:00
freq,1,1,96478,3,9,47,3,522


In [30]:
orders.isna().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [31]:
orders.duplicated().sum()

np.int64(0)

In [32]:
orders.nunique()

order_id                         99441
customer_id                      99441
order_status                         8
order_purchase_timestamp         98875
order_approved_at                90733
order_delivered_carrier_date     81018
order_delivered_customer_date    95664
order_estimated_delivery_date      459
dtype: int64

In [104]:
print("Temporal Coverage Check : Orders")
print("-" * 80)

date_cols_orders = [
    "order_purchase_timestamp",
    "order_approved_at",
    "order_delivered_carrier_date",
    "order_delivered_customer_date",
    "order_estimated_delivery_date",
]

for col in date_cols_orders:
    orders[col] = pd.to_datetime(orders[col], errors="coerce")
    print(f"{col:35s} : {orders[col].min()}  →  {orders[col].max()}")

Temporal Coverage Check : Orders
--------------------------------------------------------------------------------
order_purchase_timestamp            : 2016-09-04 21:15:19  →  2018-10-17 17:30:18
order_approved_at                   : 2016-09-15 12:16:38  →  2018-09-03 17:40:06
order_delivered_carrier_date        : 2016-10-08 10:34:01  →  2018-09-11 19:48:28
order_delivered_customer_date       : 2016-10-11 13:46:32  →  2018-10-17 13:22:46
order_estimated_delivery_date       : 2016-09-30 00:00:00  →  2018-11-12 00:00:00


In [110]:
print("Business Logic Check : Order Date Sequence")
print("-" * 55)

invalid_approved = orders.loc[orders["order_approved_at"] < orders["order_purchase_timestamp"]]
invalid_delivered_carrier = orders.loc[orders["order_delivered_carrier_date"] < orders["order_approved_at"]]
invalid_delivered_customer = orders.loc[orders["order_delivered_customer_date"] < orders["order_delivered_carrier_date"]]

print(f"approved_at < purchase_timestamp        : {len(invalid_approved):,} baris")
print(f"delivered_carrier < approved_at         : {len(invalid_delivered_carrier):,} baris")
print(f"delivered_customer < delivered_carrier  : {len(invalid_delivered_customer):,} baris")

Business Logic Check : Order Date Sequence
-------------------------------------------------------
approved_at < purchase_timestamp        : 0 baris
delivered_carrier < approved_at         : 1,359 baris
delivered_customer < delivered_carrier  : 23 baris


In [108]:
print("Business Logic Check : Order Status vs Delivery Date Consistency")
print("-" * 65)

status_summary = orders.groupby("order_status")["order_delivered_customer_date"].apply(
    lambda x: x.isna().sum()
).reset_index(name="missing_delivered_date")

print(status_summary)

Business Logic Check : Order Status vs Delivery Date Consistency
-----------------------------------------------------------------
  order_status  missing_delivered_date
0     approved                       2
1     canceled                     619
2      created                       5
3    delivered                       8
4     invoiced                     314
5   processing                     301
6      shipped                    1107
7  unavailable                     609


#### Initial Observation Orders Dataset

- Dataset terdiri dari **99,441** orders records dan **8** kolom.
- Terdapat Beberapa Kolom memiliki `NaN` di antaranya
    - `order_approved_at` sebanyak **160** baris
    - `order_delivered_carrier_date` sebanyak **1783**
    - `order_delivered_customer_date`sebanyak **2965**
- Tidak ditemukan baris duplikat.
- Terdapat 8 Kolom dengan data type `str`
- Setiap `order_id` bersifat unik dan dapat digunakan sebagai identifier utama.
- Setiap baris `order_id` memiliki `customer_id` 
- Dataset Order memiliki periode `order_purchase_timestamp` dari [04-09-2016] sampai [17-10-2018] — total rentang [25] bulan."
- Ditemukan ketidaksesuian urutan tanggal dalam dataset yaitu 
    - terdapat **1.359** baris dengan tanggal `delivered_carrier_date` yang lebih awal dari `order_approved_at` 
    - terdapat **23** baris dengan tanggal `delivered_customer_date` yang lebih awal dari `order_delivered_carrier_date`. 
Ketidaksesuaian urutan tanggal ini akan ditangani pada tahap data cleaning.
- Ditemukan **8** baris dengan status `delivered` tetapi `order_delivered_customer_date` bernilai kosong


## order items dataset

In [ ]:
order_items.sample(5)

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14


In [35]:
order_items.sample(5)

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
38381,57274bbddd47fb5bf2f1d08edb906d18,1,79366d6a24de9351b7ca6e3cf75a68ec,cc3d14116804f7209f3884f6367f1152,2018-06-27 03:17:28,193.00,24.15
34683,4e7c5bab2c0ee8406eb8aa8b2fd24d5d,1,7d50930301709838d8e90dcac5cfaa84,9616352088dcf83a7c06637f4ebf1c80,2017-08-21 02:45:59,59.90,16.86
40228,5b93109649fd0855758bb61ef5fc3d37,1,2b9496ed8ff2f7b2003258b94a58772b,7a67c85e85bb2ce8582c35f2203ad736,2018-04-16 18:09:12,129.99,23.49
46826,6a7509b21939697564bb0b5aac2c8187,1,99a4788cb24856965c36a24e339b6058,4a3ca9315b744ce9f8e9374361493884,2017-11-30 00:18:51,89.90,21.12
16558,25e7a9bc97827149a1de2b94ecdc21ef,1,422879e10f46682990de24d770e7f83d,1f50f920176fa81dab994f9023523100,2017-11-30 13:33:51,49.00,28.79


In [36]:
order_items.shape

(112650, 7)

In [37]:
order_items.info()

<class 'pandas.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  str    
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  str    
 3   seller_id            112650 non-null  str    
 4   shipping_limit_date  112650 non-null  str    
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), str(4)
memory usage: 6.0 MB


In [38]:
order_items.describe(include="all")

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
count,112650,112650.000000,112650,112650,112650,112650.000000,112650.000000
unique,98666,NaN,32951,3095,93318,NaN,NaN
top,8272b63d03f5f79c56e9e4120aec44ef,NaN,aca2eb7d00ea1a7b8ebd4e68314663af,6560211a19b47992c3666cc44a7e94c0,2018-03-01 02:50:48,NaN,NaN
freq,21,NaN,527,2033,21,NaN,NaN
mean,NaN,1.197834,NaN,NaN,NaN,120.653739,19.990320
std,NaN,0.705124,NaN,NaN,NaN,183.633928,15.806405
min,NaN,1.000000,NaN,NaN,NaN,0.850000,0.000000
25%,NaN,1.000000,NaN,NaN,NaN,39.900000,13.080000
50%,NaN,1.000000,NaN,NaN,NaN,74.990000,16.260000
75%,NaN,1.000000,NaN,NaN,NaN,134.900000,21.150000


In [39]:
order_items.isna().sum()

order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

In [40]:
order_items.duplicated().sum()

np.int64(0)

In [ ]:
order_items.nunique()

order_id               98666
order_item_id             21
product_id             32951
seller_id               3095
shipping_limit_date    93318
price                   5968
freight_value           6999
dtype: int64

In [116]:
print("=" * 70)
print("Impossible Value Check : Order Items")
print("=" * 70)

invalid_price = order_items.loc[order_items["price"] <= 0]
invalid_freight = order_items.loc[order_items["freight_value"] < 0]

print(f"price <= 0            : {len(invalid_price):,} baris")
print(f"freight_value < 0     : {len(invalid_freight):,} baris")

Impossible Value Check : Order Items
price <= 0            : 0 baris
freight_value < 0     : 0 baris


#### Initial Observation Order Item Dataset

- Dataset terdiri dari **112,650** Order item records dan **7** kolom.
- Tidak ditemukan missing values pada seluruh kolom.
- Tidak ditemukan baris duplikat.
- Terdapat 4 Kolom dengan data type `str`, 1 Kolom dengan data type `int64` dan 2 Kolom dengan data type `float64`
- Setiap kolom order item memiliki beberapa unique id dari beberapa tabel didalam nya yaitu `order_id`, `order_item_id`, `product_id` dan `seller_id`
- Tidak ditemukan nilai `price` yang kurang dari atau sama dengan 0, sehingga seluruh item memiliki harga yang valid secara numerik.
- Tidak ditemukan nilai `freight_value` negatif, sehingga tidak terdapat biaya pengiriman dengan nilai yang secara numerik tidak valid.

## Payments Dataset

In [43]:
payments.head()

,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71
3,ba78997921bbcdc1373bb41e913ab953,1,credit_card,8,107.78
4,42fdf880ba16b47b59251dd489d4441a,1,credit_card,2,128.45


In [44]:
payments.sample(5)

,order_id,payment_sequential,payment_type,payment_installments,payment_value
23913,8cf12a7b163c4225da01e9da10cce6fb,1,credit_card,2,46.10
19454,942e1205a3ca6fd0147718ac16f4bb54,1,boleto,1,193.19
80394,4db4d9eaf5fb240987ed4148caf900b9,1,credit_card,1,36.28
43729,49fa1c392707716cd5d12cc1c7c65c15,1,credit_card,1,260.64
85926,f9556cd22578140995b4179fffee7c4a,1,credit_card,5,93.50


In [45]:
payments.shape

(103886, 5)

In [46]:
payments.info()

<class 'pandas.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  str    
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  str    
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), str(2)
memory usage: 4.0 MB


In [48]:
payments.describe(include="all")

,order_id,payment_sequential,payment_type,payment_installments,payment_value
count,103886,103886.000000,103886,103886.000000,103886.000000
unique,99440,NaN,5,NaN,NaN
top,fa65dad1b0e818e3ccc5cb0e39231352,NaN,credit_card,NaN,NaN
freq,29,NaN,76795,NaN,NaN
mean,NaN,1.092679,NaN,2.853349,154.100380
std,NaN,0.706584,NaN,2.687051,217.494064
min,NaN,1.000000,NaN,0.000000,0.000000
25%,NaN,1.000000,NaN,1.000000,56.790000
50%,NaN,1.000000,NaN,1.000000,100.000000
75%,NaN,1.000000,NaN,4.000000,171.837500


In [49]:
payments.isna().sum()

order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64

In [51]:
payments.duplicated().sum()

np.int64(0)

In [52]:
payments.nunique()

order_id                99440
payment_sequential         29
payment_type                5
payment_installments       24
payment_value           29077
dtype: int64

In [117]:
print("=" * 70)
print("Impossible Value Check : Payments")
print("=" * 70)

invalid_payment_value = payments.loc[payments["payment_value"] <= 0]
invalid_installments = payments.loc[payments["payment_installments"] <= 0]

print(f"payment_value <= 0        : {len(invalid_payment_value):,} baris")
print(f"payment_installments <= 0 : {len(invalid_installments):,} baris")

Impossible Value Check : Payments
payment_value <= 0        : 9 baris
payment_installments <= 0 : 2 baris


#### Initial Observation Payment Dataset

- Dataset terdiri dari **103.886** customer records dan **5** kolom.
- Tidak ditemukan missing values pada seluruh kolom.
- Tidak ditemukan baris duplikat.
- Terdapat 2 Kolom dengan data type `str`, 2 Kolom dengan data type `int64` dan 1 kolom dengan data type `float64`
- Setiap baris payment memiliki `order_id` sebagai identifier 
- Ditemukan **9** baris dengan `payment_value` kurang dari atau sama dengan 0, yang menunjukkan adanya nilai pembayaran yang tidak valid secara numerik.
- Ditemukan **2** baris dengan `payment_installments` kurang dari atau sama dengan 0, yang menunjukkan adanya jumlah cicilan yang tidak valid.
- Total terdapat **11** baris yang memiliki nilai payment tidak sesuai dengan business logic dan perlu ditinjau pada tahap data cleaning.

## Products Dataset 

In [54]:
products.head()

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0
3,cef67bcfe19066a932b7673e239eb23d,bebes,27.0,261.0,1.0,371.0,26.0,4.0,26.0
4,9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37.0,402.0,4.0,625.0,20.0,17.0,13.0


In [55]:
products.sample(5)

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
27632,87716aafd7c6269cddb3403f85e4102e,utilidades_domesticas,32.0,84.0,2.0,515.0,39.0,11.0,25.0
29280,3c1fe640b9b303b8f12b612e350b9446,brinquedos,29.0,556.0,2.0,3075.0,44.0,45.0,23.0
32254,fddc61a36e83f12b665d068c07411aa7,esporte_lazer,50.0,2164.0,6.0,975.0,33.0,30.0,28.0
14026,a84e26c4cc7ea7884d3a81955a06cefc,papelaria,54.0,726.0,6.0,800.0,30.0,3.0,22.0
16273,f7da48a5b38c79d4f161133afca9ebc1,informatica_acessorios,37.0,1319.0,2.0,417.0,20.0,13.0,16.0


In [56]:
products.shape

(32951, 9)

In [57]:
products.info()

<class 'pandas.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  str    
 1   product_category_name       32341 non-null  str    
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), str(2)
memory usage: 2.3 MB


In [58]:
products.describe(include="all")

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
count,32951,32341,32341.000000,32341.000000,32341.000000,32949.000000,32949.000000,32949.000000,32949.000000
unique,32951,73,NaN,NaN,NaN,NaN,NaN,NaN,NaN
top,1e9e8ef04dbcff4541ed26657ea517e5,cama_mesa_banho,NaN,NaN,NaN,NaN,NaN,NaN,NaN
freq,1,3029,NaN,NaN,NaN,NaN,NaN,NaN,NaN
mean,NaN,NaN,48.476949,771.495285,2.188986,2276.472488,30.815078,16.937661,23.196728
std,NaN,NaN,10.245741,635.115225,1.736766,4282.038731,16.914458,13.637554,12.079047
min,NaN,NaN,5.000000,4.000000,1.000000,0.000000,7.000000,2.000000,6.000000
25%,NaN,NaN,42.000000,339.000000,1.000000,300.000000,18.000000,8.000000,15.000000
50%,NaN,NaN,51.000000,595.000000,1.000000,700.000000,25.000000,13.000000,20.000000
75%,NaN,NaN,57.000000,972.000000,3.000000,1900.000000,38.000000,21.000000,30.000000


In [59]:
products.isna().sum()

product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

In [60]:
products.duplicated().sum()

np.int64(0)

In [61]:
products.nunique()

product_id                    32951
product_category_name            73
product_name_lenght              66
product_description_lenght     2960
product_photos_qty               19
product_weight_g               2204
product_length_cm                99
product_height_cm               102
product_width_cm                 95
dtype: int64

#### Initial Observation Products Dataset

- Dataset terdiri dari **32.951** products records dan **9** kolom.
- Beberapa Kolom terdapat Missing Values yaitu,
    - `product_category_name` sebanyak **610** baris
    - `product_name_lenght`sebanyak **610** baris
    - `product_description_lenght`sebanyak **610** baris
    - `product_photos_qty`sebanyak **610** baris
- Tidak ditemukan baris duplikat.
- Terdapat 2 Kolom dengan data type `str` dan 7 Kolom dengan data type `float64`
- Setiap `product_id` bersifat unik dan dapat digunakan sebagai identifier utama.

## Reviews Datsaset

In [62]:
reviews.head()

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,2018-01-18 00:00:00,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,2018-03-10 00:00:00,2018-03-11 03:05:13
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,NaN,NaN,2018-02-17 00:00:00,2018-02-18 14:36:24
3,e64fb393e7b32834bb789ff8bb30750e,658677c97b385a9be170737859d3511b,5,NaN,Recebi bem antes do prazo estipulado.,2017-04-21 00:00:00,2017-04-21 22:02:06
4,f7c4243c7fe1938f181bec41a392bdeb,8e6bfb81e283fa7e4f11123a3fb894f1,5,NaN,Parabéns lojas lannister adorei comprar pela I...,2018-03-01 00:00:00,2018-03-02 10:26:53


In [63]:
reviews.sample(5)

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
22199,644ae28209f51f17b1d71bab566b4149,c6ee16bc69a830e4c2e978c559071137,5,NaN,Bom,2017-08-29 00:00:00,2017-09-01 12:38:12
35563,ffaf83e3350ca02b32c119c2bcf98ec8,cad95e56b8bc5ed2c10ee964ba21ac96,2,NaN,NaN,2018-03-21 00:00:00,2018-03-21 12:52:32
64655,c36b2af48975e0f1ee2c6c1405ceb560,59754064fe4ffec6c0149904a6573d81,1,NaN,NaN,2017-08-06 00:00:00,2017-08-12 11:03:52
62423,59ae784a4b793e1277ed02f6aa85135b,2fe2b3f315aee60c3f8f15af8d3e0ab6,5,NaN,NaN,2018-06-14 00:00:00,2018-06-14 20:42:38
12002,47d29634c206c8dc4c73221b31e6004e,40bc2b2661f13c1a9888396ed56ba79a,5,NaN,NaN,2018-05-04 00:00:00,2018-05-08 01:14:16


In [64]:
reviews.shape

(99224, 7)

In [65]:
reviews.info()

<class 'pandas.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype
---  ------                   --------------  -----
 0   review_id                99224 non-null  str  
 1   order_id                 99224 non-null  str  
 2   review_score             99224 non-null  int64
 3   review_comment_title     11568 non-null  str  
 4   review_comment_message   40977 non-null  str  
 5   review_creation_date     99224 non-null  str  
 6   review_answer_timestamp  99224 non-null  str  
dtypes: int64(1), str(6)
memory usage: 5.3 MB


In [66]:
reviews.describe(include="all")

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
count,99224,99224,99224.000000,11568,40977,99224,99224
unique,98410,98673,NaN,4527,36159,636,98248
top,c444278834184f72b1484dfe47de7f97,c88b1d1b157a9999ce368f218a407141,NaN,Recomendo,Muito bom,2017-12-19 00:00:00,2017-06-15 23:21:05
freq,3,3,NaN,423,230,463,4
mean,NaN,NaN,4.086421,NaN,NaN,NaN,NaN
std,NaN,NaN,1.347579,NaN,NaN,NaN,NaN
min,NaN,NaN,1.000000,NaN,NaN,NaN,NaN
25%,NaN,NaN,4.000000,NaN,NaN,NaN,NaN
50%,NaN,NaN,5.000000,NaN,NaN,NaN,NaN
75%,NaN,NaN,5.000000,NaN,NaN,NaN,NaN


In [67]:
reviews.isna().sum()

review_id                      0
order_id                       0
review_score                   0
review_comment_title       87656
review_comment_message     58247
review_creation_date           0
review_answer_timestamp        0
dtype: int64

In [68]:
reviews.duplicated().sum()

np.int64(0)

In [69]:
reviews.nunique()

review_id                  98410
order_id                   98673
review_score                   5
review_comment_title        4527
review_comment_message     36159
review_creation_date         636
review_answer_timestamp    98248
dtype: int64

In [112]:
print("Temporal Coverage Check : Reviews")
print("-" * 80)

reviews["review_creation_date"] = pd.to_datetime(reviews["review_creation_date"], errors="coerce")
reviews["review_answer_timestamp"] = pd.to_datetime(reviews["review_answer_timestamp"], errors="coerce")

print(f"review_creation_date               : {reviews['review_creation_date'].min()}  →  {reviews['review_creation_date'].max()}")
print(f"review_answer_timestamp            : {reviews['review_answer_timestamp'].min()}  →  {reviews['review_answer_timestamp'].max()}")

Temporal Coverage Check : Reviews
--------------------------------------------------------------------------------
review_creation_date               : 2016-10-02 00:00:00  →  2018-08-31 00:00:00
review_answer_timestamp            : 2016-10-07 18:32:28  →  2018-10-29 12:27:35


In [114]:
print("Business Logic Check : Review Date Sequence")
print("-" * 50)

invalid_review = reviews.loc[reviews["review_answer_timestamp"] < reviews["review_creation_date"]]
print(f"answer_timestamp < creation_date : {len(invalid_review):,} baris")

Business Logic Check : Review Date Sequence
--------------------------------------------------
answer_timestamp < creation_date : 0 baris


In [115]:
print("Impossible Value Check : Reviews")
print("-" * 70)

invalid_score = reviews.loc[~reviews["review_score"].between(1, 5)]
print(f"review_score di luar rentang 1-5 : {len(invalid_score):,} baris")

Impossible Value Check : Reviews
----------------------------------------------------------------------
review_score di luar rentang 1-5 : 0 baris


#### Initial Observation Reviews Dataset

- Dataset terdiri dari **99.224** reviews records dan **7** kolom.
- Terdapat 2 kolom dari dataset review yang memiliki banyak Missing Values yaitu, 
    - `review_comment_title` sebanyak **87.656** baris
    - `review_comment_message` sebanyak **58.247** baris
- Tidak ditemukan baris duplikat.
- Terdapat 6 Kolom dengan data type `str` dan 1 Kolom dengan data type `int64`
- Setiap `review_id` bersifat unik dan dapat digunakan sebagai identifier utama, setiap reviews memiliki `order_id`
- Periode `review_creation_date` mencakup [02-10-2016] hingga [31-08-2018] sedangkan `review_answer_timestamp` mencakup [07-10-2016] hingga [29-10-2018]
- Tidak ditemukan ketidaksesuaian urutan waktu antara `review_creation_date` dan `review_answer_timestamp`, dengan 0 baris di mana review dijawab sebelum review dibuat.
- Seluruh nilai review_score berada dalam rentang 1–5, sehingga tidak ditemukan nilai yang tidak valid pada variabel skor review.

## Sellers Dataset

In [70]:
sellers.head()

,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ
3,c0f3eea2e14555b6faeea3dd58c1b1c3,4195,sao paulo,SP
4,51a04a8a6bdcb23deccc82b0b80742cf,12914,braganca paulista,SP


In [74]:
sellers.sample(5)

,seller_id,seller_zip_code_prefix,seller_city,seller_state
406,48e954b2da4f6a750c844173b347ed40,91740,porto alegre,RS
2568,35857757f553273b1056c1cabdace7fb,14020,ribeirao preto,SP
1753,f3511c85f59f8dec53d140501ee8e7a8,80045,curitiba,PR
1375,28ea411b354f14b1a494be4e79f0ce02,5685,sao paulo,SP
837,02c988090b766852e088c69d7fb3b551,13473,americana,SP


In [72]:
sellers.shape

(3095, 4)

In [75]:
sellers.info()

<class 'pandas.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype
---  ------                  --------------  -----
 0   seller_id               3095 non-null   str  
 1   seller_zip_code_prefix  3095 non-null   int64
 2   seller_city             3095 non-null   str  
 3   seller_state            3095 non-null   str  
dtypes: int64(1), str(3)
memory usage: 96.8 KB


In [76]:
sellers.describe(include="all")

,seller_id,seller_zip_code_prefix,seller_city,seller_state
count,3095,3095.000000,3095,3095
unique,3095,NaN,611,23
top,3442f8959a84dea7ee197c632cb2df15,NaN,sao paulo,SP
freq,1,NaN,694,1849
mean,NaN,32291.059451,NaN,NaN
std,NaN,32713.453830,NaN,NaN
min,NaN,1001.000000,NaN,NaN
25%,NaN,7093.500000,NaN,NaN
50%,NaN,14940.000000,NaN,NaN
75%,NaN,64552.500000,NaN,NaN


In [77]:
sellers.isna().sum()

seller_id                 0
seller_zip_code_prefix    0
seller_city               0
seller_state              0
dtype: int64

In [78]:
sellers.duplicated().sum()

np.int64(0)

In [79]:
sellers.nunique()

seller_id                 3095
seller_zip_code_prefix    2246
seller_city                611
seller_state                23
dtype: int64

#### Initial Observation Sellers Dataset

- Dataset terdiri dari **3.095** customer records dan **4** kolom.
- Tidak ditemukan missing values pada seluruh kolom.
- Tidak ditemukan baris duplikat.
- Terdapat 3 Kolom dengan data type `str` dan 1 Kolom dengan data type `int64`
- Setiap `seller_id` bersifat unik dan dapat digunakan sebagai identifier utama.

## Geolocation Dataset

In [80]:
geo.head()

,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.545621,-46.639292,sao paulo,SP
1,1046,-23.546081,-46.644820,sao paulo,SP
2,1046,-23.546129,-46.642951,sao paulo,SP
3,1041,-23.544392,-46.639499,sao paulo,SP
4,1035,-23.541578,-46.641607,sao paulo,SP


In [81]:
geo.sample(5)

,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
775951,69038,-3.080938,-60.066755,manaus,AM
224059,9210,-23.635568,-46.534833,santo andre,SP
93887,4569,-23.612036,-46.690045,sao paulo,SP
290825,13160,-22.579302,-47.175462,artur nogueira,SP
944651,90660,-30.060100,-51.196311,porto alegre,RS


In [82]:
geo.shape

(1000163, 5)

In [83]:
geo.info()

<class 'pandas.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  int64  
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_lng              1000163 non-null  float64
 3   geolocation_city             1000163 non-null  str    
 4   geolocation_state            1000163 non-null  str    
dtypes: float64(2), int64(1), str(2)
memory usage: 38.2 MB


In [84]:
geo.describe(include="all")

,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
count,1.000163e+06,1.000163e+06,1.000163e+06,1000163,1000163
unique,NaN,NaN,NaN,8011,27
top,NaN,NaN,NaN,sao paulo,SP
freq,NaN,NaN,NaN,135800,404268
mean,3.657417e+04,-2.117615e+01,-4.639054e+01,NaN,NaN
std,3.054934e+04,5.715866e+00,4.269748e+00,NaN,NaN
min,1.001000e+03,-3.660537e+01,-1.014668e+02,NaN,NaN
25%,1.107500e+04,-2.360355e+01,-4.857317e+01,NaN,NaN
50%,2.653000e+04,-2.291938e+01,-4.663788e+01,NaN,NaN
75%,6.350400e+04,-1.997962e+01,-4.376771e+01,NaN,NaN


In [85]:
geo.isna().sum()

geolocation_zip_code_prefix    0
geolocation_lat                0
geolocation_lng                0
geolocation_city               0
geolocation_state              0
dtype: int64

In [86]:
geo.duplicated().sum()

np.int64(261831)

In [87]:
geo.nunique()

geolocation_zip_code_prefix     19015
geolocation_lat                717363
geolocation_lng                717615
geolocation_city                 8011
geolocation_state                  27
dtype: int64

#### Initial Observation Geoloaction Dataset

- Dataset terdiri dari **1.000.163** customer records dan **5** kolom.
- Tidak ditemukan missing values pada seluruh kolom.
- Tidak ditemukan baris duplikat.
- Terdapat 2 Kolom dengan data type `str`, 2 Kolom dengan data type `float64` dam 1 kolom dengan data type `int64`
- Setiap kolom menjelaskan posisi lokasi di lengkapi dengan zip code, latitude, longtitude, city dan state

## Category Name Translation Dataset 

In [88]:
translation.head()

,product_category_name,product_category_name_english
0,beleza_saude,health_beauty
1,informatica_acessorios,computers_accessories
2,automotivo,auto
3,cama_mesa_banho,bed_bath_table
4,moveis_decoracao,furniture_decor


In [89]:
translation.sample(5)

,product_category_name,product_category_name_english
43,construcao_ferramentas_iluminacao,construction_tools_lights
35,casa_conforto,home_confort
33,eletrodomesticos,home_appliances
10,alimentos_bebidas,food_drink
27,construcao_ferramentas_jardim,costruction_tools_garden


In [90]:
translation.shape

(71, 2)

In [91]:
translation.info()

<class 'pandas.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 2 columns):
 #   Column                         Non-Null Count  Dtype
---  ------                         --------------  -----
 0   product_category_name          71 non-null     str  
 1   product_category_name_english  71 non-null     str  
dtypes: str(2)
memory usage: 1.2 KB


In [92]:
translation.describe(include="all")

,product_category_name,product_category_name_english
count,71,71
unique,71,71
top,beleza_saude,health_beauty
freq,1,1


In [93]:
translation.isna().sum()

product_category_name            0
product_category_name_english    0
dtype: int64

In [94]:
translation.duplicated().sum()

np.int64(0)

In [95]:
translation.nunique()

product_category_name            71
product_category_name_english    71
dtype: int64

#### Initial Observation Category Name Translation Dataset

- Dataset terdiri dari **71** customer records dan **2** kolom.
- Tidak ditemukan missing values pada seluruh kolom.
- Tidak ditemukan baris duplikat.
- Terdapat 2 Kolom dengan data type `str` 
- Setiap `product_category_name` memiliki kolom `product_category_name_english`

## Cross-Dataset Relationship Check

### Objective

Memverifikasi bahwa hubungan antar dataset konsisten berdasarkan key yang dimiliki masing-masing tabel.

### Output

- Validasi hubungan antar tabel.
- Jumlah foreign key yang tidak memiliki pasangan (jika ada).
- Status setiap relationship (PASSED / FAILED).

### Notes

Tahap ini hanya melakukan pengecekan keberadaan foreign key dan belum melakukan validasi bisnis secara mendalam.

### Customers → Orders

**Objective**

Memastikan setiap `customer_id` pada tabel **orders** memiliki pasangan pada tabel **customers**.

In [96]:
missing_customers = orders.loc[
    ~orders["customer_id"].isin(customers["customer_id"])
]

print("=" * 70)
print("Relationship Check : Customers → Orders")
print("=" * 70)

print(f"Total Orders                 : {len(orders):,}")
print(f"Orders without Customer      : {len(missing_customers):,}")

if len(missing_customers) == 0:
    print("Relationship Status          : PASSED ✅")
else:
    print("Relationship Status          : FAILED ❌")

Relationship Check : Customers → Orders
Total Orders                 : 99,441
Orders without Customer      : 0
Relationship Status          : PASSED ✅


### Orders → Order Items

**Objective**

Memastikan seluruh `order_id` pada tabel **order_items** memiliki pasangan pada tabel **orders**.

In [97]:
missing_orders = order_items.loc[
    ~order_items["order_id"].isin(orders["order_id"])
]

print("=" * 70)
print("Relationship Check : Orders → Order Items")
print("=" * 70)

print(f"Total Order Items           : {len(order_items):,}")
print(f"Missing Orders              : {len(missing_orders):,}")

if len(missing_orders) == 0:
    print("Relationship Status         : PASSED ✅")
else:
    print("Relationship Status         : FAILED ❌")

Relationship Check : Orders → Order Items
Total Order Items           : 112,650
Missing Orders              : 0
Relationship Status         : PASSED ✅


### Orders → Payments

**Objective**

Memastikan seluruh transaksi pembayaran (`order_id`) memiliki pasangan pada tabel **orders**.

In [98]:
missing_payments = payments.loc[
    ~payments["order_id"].isin(orders["order_id"])
]

print("=" * 70)
print("Relationship Check : Orders → Payments")
print("=" * 70)

print(f"Total Payments              : {len(payments):,}")
print(f"Missing Orders              : {len(missing_payments):,}")

if len(missing_payments) == 0:
    print("Relationship Status         : PASSED ✅")
else:
    print("Relationship Status         : FAILED ❌")

Relationship Check : Orders → Payments
Total Payments              : 103,886
Missing Orders              : 0
Relationship Status         : PASSED ✅


### Orders → Reviews

**Objective**

Memastikan seluruh ulasan (`order_id`) memiliki pasangan pada tabel **orders**.

In [99]:
missing_reviews = reviews.loc[
    ~reviews["order_id"].isin(orders["order_id"])
]

print("=" * 70)
print("Relationship Check : Orders → Reviews")
print("=" * 70)

print(f"Total Reviews               : {len(reviews):,}")
print(f"Missing Orders              : {len(missing_reviews):,}")

if len(missing_reviews) == 0:
    print("Relationship Status         : PASSED ✅")
else:
    print("Relationship Status         : FAILED ❌")

Relationship Check : Orders → Reviews
Total Reviews               : 99,224
Missing Orders              : 0
Relationship Status         : PASSED ✅


### Order Items → Products

**Objective**

Memastikan seluruh `product_id` pada tabel **order_items** tersedia pada tabel **products**.

In [100]:
missing_products = order_items.loc[
    ~order_items["product_id"].isin(products["product_id"])
]

print("=" * 70)
print("Relationship Check : Order Items → Products")
print("=" * 70)

print(f"Total Order Items           : {len(order_items):,}")
print(f"Missing Products            : {len(missing_products):,}")

if len(missing_products) == 0:
    print("Relationship Status         : PASSED ✅")
else:
    print("Relationship Status         : FAILED ❌")

Relationship Check : Order Items → Products
Total Order Items           : 112,650
Missing Products            : 0
Relationship Status         : PASSED ✅


### Order Items → Sellers

**Objective**

Memastikan seluruh `seller_id` pada tabel **order_items** tersedia pada tabel **sellers**.

In [101]:
missing_sellers = order_items.loc[
    ~order_items["seller_id"].isin(sellers["seller_id"])
]

print("=" * 70)
print("Relationship Check : Order Items → Sellers")
print("=" * 70)

print(f"Total Order Items           : {len(order_items):,}")
print(f"Missing Sellers             : {len(missing_sellers):,}")

if len(missing_sellers) == 0:
    print("Relationship Status         : PASSED ✅")
else:
    print("Relationship Status         : FAILED ❌")

Relationship Check : Order Items → Sellers
Total Order Items           : 112,650
Missing Sellers             : 0
Relationship Status         : PASSED ✅


## Initial Dataset Summary

### Objective

Merangkum hasil utama dari proses data profiling sebagai dasar untuk tahap **Dataset Acquisition & Assessment**.

Bagian ini hanya mencatat kondisi awal dataset berdasarkan hasil profiling dan validasi relationship. Belum dilakukan analisis bisnis maupun penarikan insight.

### Output

- Ringkasan kondisi dataset.
- Temuan teknis awal.
- Kesiapan dataset untuk tahap selanjutnya.

### Notes

Kesimpulan pada bagian ini bersifat teknis dan akan menjadi referensi dalam proses Dataset Assessment serta Database Design.

## Additional Data Quality Findings

Bagian ini merangkum temuan tambahan dari validasi cakupan temporal, konsistensi logika bisnis, dan nilai mustahil yang dilakukan lintas tabel selama proses profiling. Temuan ini menjadi acuan untuk keputusan cleaning pada tahap berikutnya dan akan dirujuk kembali pada `data_quality_report.md`.

### Identifier Consistency
- `customer_id` (Customers/Orders) bersifat unik per baris (identifier level order), sedangkan `customer_unique_id` TIDAK unik (96,096 unique dari 99,441 baris) — terdapat 3,345 baris dengan pelanggan yang order lebih dari sekali.
- **Keputusan yang perlu diambil:** seluruh analisis customer-level (segmentation, churn, RFM) wajib menggunakan `customer_unique_id`, bukan `customer_id`.

### Temporal Coverage
| Tabel | Kolom | Rentang |
|---|---|---|
| Orders | `order_purchase_timestamp` | 04-09-2016 → 17-10-2018 (~25 bulan) |
| Orders | `order_delivered_customer_date` | 11-10-2016 → 17-10-2018 |
| Reviews | `review_creation_date` | 02-10-2016 → 31-08-2018 |
| Reviews | `review_answer_timestamp` | 07-10-2016 → 29-10-2018 |

Catatan: total cakupan data ±25 bulan. Ini menjadi batas atas untuk validasi threshold churn 180 hari — periode observasi tidak cukup panjang untuk mengukur siklus churn di atas ~12–18 bulan dengan keyakinan tinggi.

### Business Logic Anomalies
| Tabel | Pemeriksaan | Jumlah Baris Anomali |
|---|---|---|
| Orders | `order_delivered_carrier_date` < `order_approved_at` | 1,359 |
| Orders | `order_delivered_customer_date` < `order_delivered_carrier_date` | 23 |
| Orders | status `delivered` tapi `order_delivered_customer_date` kosong | 8 |
| Reviews | `review_answer_timestamp` < `review_creation_date` | 0 |

**Keputusan yang perlu diambil:** 1,382 baris Orders (1,359 + 23) memiliki urutan tanggal yang tidak logis dan perlu ditinjau — apakah di-drop, dikoreksi, atau dibiarkan dengan flag khusus saat tahap cleaning nanti.

### Impossible Values
| Tabel | Pemeriksaan | Jumlah Baris Anomali |
|---|---|---|
| Order Items | `price` ≤ 0 | 0 |
| Order Items | `freight_value` < 0 | 0 |
| Payments | `payment_value` ≤ 0 | 9 |
| Payments | `payment_installments` ≤ 0 | 2 |
| Reviews | `review_score` di luar rentang 1–5 | 0 |

**Keputusan yang perlu diambil:** 11 baris Payments dengan nilai tidak valid perlu ditinjau — kemungkinan besar terkait transaksi yang dibatalkan/gratis, perlu dicek silang dengan `order_status` sebelum diputuskan di-drop atau tidak.

### Summary
Total baris bermasalah dari seluruh dataset (Orders + Payments): **1,393 baris** dari total ~99,441 order records (~1.4%). Skala ini tergolong kecil dan tidak menghambat kelayakan dataset secara keseluruhan, namun tetap wajib dicatat sebagai keputusan eksplisit di Decision Log sebelum data cleaning dimulai.

## Observation

Berdasarkan proses data profiling, seluruh sembilan dataset berhasil dimuat dan diperiksa tanpa kendala. Struktur dataset menunjukkan bahwa Olist merupakan dataset relasional yang dibangun menggunakan beberapa key utama seperti `order_id`, `product_id`, dan `seller_id`. Validasi relationship antar tabel menunjukkan bahwa foreign key utama berhasil dipetakan dengan benar (seluruh cross-table check PASSED), sehingga dataset memiliki integritas relasional yang baik.

Satu temuan penting: `customer_id` ternyata bersifat unik per baris (level order), sedangkan identifier pelanggan sesungguhnya adalah `customer_unique_id`, yang menunjukkan 3,345 baris pelanggan dengan lebih dari satu order. Temuan ini krusial karena seluruh analisis customer-level pada tahap berikutnya (segmentation, churn) wajib menggunakan `customer_unique_id`.

Validasi cakupan temporal menunjukkan data mencakup periode ±25 bulan (September 2016 – Oktober 2018), yang akan menjadi acuan dalam memvalidasi asumsi threshold churn 180 hari pada BRD. Validasi logika bisnis menemukan 1,382 baris pada tabel Orders dengan urutan tanggal yang tidak konsisten, serta 11 baris pada tabel Payments dengan nilai numerik yang tidak valid (payment_value atau installments ≤ 0) — totalnya sekitar 1.4% dari total order records. Kolom timestamp pada beberapa tabel juga masih bertipe `object` dan perlu dikonversi ke `datetime` pada tahap ETL.

Secara keseluruhan, tidak ditemukan permasalahan struktural yang menghambat kelayakan dataset. Anomali yang ditemukan berskala kecil dan sudah teridentifikasi dengan jelas, sehingga dataset dinilai layak untuk memasuki tahap **Dataset Assessment**, dengan catatan bahwa keputusan penanganan anomali (baris dengan urutan tanggal tidak logis, nilai payment tidak valid, dan penggunaan `customer_unique_id`) perlu dicatat secara eksplisit di Decision Log sebelum data cleaning dimulai.